File: leetcode/algorithms/543_diameter_of_binary_tree/main.java
class DiameterOfBinaryTree {
    class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    /**
     * Recursion: DFS
     * - Time Complexity: O(N)
     * - Space Complexity: O(H)
     */
    private int diameter = 0;

    private int getMaxDiameter(TreeNode node) {
        if (node == null) {
            return 0;
        }

        int leftDiameter = getMaxDiameter(node.left);
        int rightDiameter = getMaxDiameter(node.right);

        if (leftDiameter + rightDiameter > diameter) {
            diameter = leftDiameter + rightDiameter;
        }

        return Math.max(leftDiameter, rightDiameter) + 1;
    }

    public int diameterOfBinaryTree(TreeNode root) {
        getMaxDiameter(root);

        return diameter;
    }


    // Solution
    /**
     * Recursion: DFS
     * - Time Complexity: O(N)
     * - Space Complexity: O(H)
     */
    private int maxDiameter = 0;

    public int Solution(TreeNode root) {
        calculateDepth(root);
        return maxDiameter;
    }

    private int calculateDepth(TreeNode node) {
        if (node == null) {
            return 0;
        }

        int leftDepth = calculateDepth(node.left);
        int rightDepth = calculateDepth(node.right);

        maxDiameter = Math.max(maxDiameter, leftDepth + rightDepth);

        return Math.max(leftDepth, rightDepth) + 1;
    }
}

File: leetcode/algorithms/543_diameter_of_binary_tree/main.js
function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}

/**
 * Recursion: DFS (Bottom-up)
 * - Time Complexity: O(N)
 * - Space Complexity: O(H)
 *
 * @param {TreeNode} root
 * @return {number}
 */
var diameterOfBinaryTree = function (root) {
  let maxDiameter = 0;

  /**
   * @param {TreeNode} node
   * @returns {number}
   */
  function getMaxDiameter(node) {
    if (node === null) {
      return 0;
    }

    const leftDiameter = getMaxDiameter(node.left);
    const rightDiameter = getMaxDiameter(node.right);

    if (leftDiameter + rightDiameter > maxDiameter) {
      maxDiameter = leftDiameter + rightDiameter;
    }

    return Math.max(leftDiameter, rightDiameter) + 1;
  }

  getMaxDiameter(root);

  return maxDiameter;
};


// Solution
/**
 * Recursion: DFS (Bottom-up)
 * - Time Complexity: O(N)
 * - Space Complexity: O(H)
 */
var solution = function (root) {
  let maxDiameter = 0;

  function dfs(node) {
    if (node === null) {
      return 0;
    }

    const leftHeight = dfs(node.left);
    const rightHeight = dfs(node.right);

    maxDiameter = Math.max(maxDiameter, leftHeight + rightHeight);

    return 1 + Math.max(leftHeight, rightHeight);
  }

  dfs(root);

  return maxDiameter;
};
